File: src/com/aopphp/go/pointcut/SignaturePointcut.java
package com.aopphp.go.pointcut;

import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.PhpClassMember;
import com.jetbrains.php.lang.psi.elements.PhpNamedElement;
import org.jetbrains.annotations.NotNull;

import java.util.Set;
import java.util.regex.Pattern;

public class SignaturePointcut implements Pointcut
{
    private final Set<KindFilter> filterKind;
    private final String name;
    private final Pattern regexp;
    private final PointFilter modifierFilter;

    private PointFilter classFilter = null;

    public SignaturePointcut(Set<KindFilter> filterKind, @NotNull String name, PointFilter modifierFilter) {
        this.filterKind = filterKind;
        this.name       = name;
        // http://stackoverflow.com/questions/14134558/
        String regexp   = name.replaceAll("([\\\\\\.\\[\\{\\(\\*\\+\\?\\^\\$\\|])", "\\\\$1");
        this.regexp     = Pattern.compile(
            regexp
                .replace("\\*\\*", ".+?")
                .replace("\\*", "[^\\\\]+?")
                .replace("\\?", ".")
                .replace("\\|", "|"));
        this.modifierFilter = modifierFilter;
    }

    @Override
    public PointFilter getClassFilter() {
        return classFilter;
    }

    public void setClassFilter(PointFilter classFilter) {
        this.classFilter = classFilter;
    }

    @Override
    public boolean matches(PhpNamedElement element) {

        if (element instanceof PhpClassMember && !modifierFilter.matches(element)) {
            return false;
        }

        String elementName = "";
        if (element instanceof PhpClassMember) {
            elementName = element.getName();
        } else if (element instanceof PhpClass) {
            String elementFQN = element.getFQN();
            elementName = elementFQN != null ? elementFQN.substring(1) : "";
        }

        return elementName.equals(name) || regexp.matcher(elementName).matches();
    }

    @Override
    public Set<KindFilter> getKind() {
        return filterKind;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof SignaturePointcut)) return false;

        SignaturePointcut that = (SignaturePointcut) o;

        if (!filterKind.equals(that.filterKind)) return false;
        if (!name.equals(that.name)) return false;
        if (!modifierFilter.equals(that.modifierFilter)) return false;
        return classFilter != null ? classFilter.equals(that.classFilter) : that.classFilter == null;
    }

    @Override
    public int hashCode() {
        int result = filterKind.hashCode();
        result = 31 * result + name.hashCode();
        result = 31 * result + modifierFilter.hashCode();
        result = 31 * result + (classFilter != null ? classFilter.hashCode() : 0);
        return result;
    }
}

File: src/main/kotlin/com/aopphp/go/pointcut/SignaturePointcut.kt
package com.aopphp.go.pointcut

import com.jetbrains.php.lang.psi.elements.PhpClass
import com.jetbrains.php.lang.psi.elements.PhpClassMember
import com.jetbrains.php.lang.psi.elements.PhpNamedElement
import java.util.regex.Pattern

class SignaturePointcut(
    private val filterKind: Set<KindFilter>,
    private val name: String,
    private val modifierFilter: PointFilter
) : Pointcut {

    private val regexp: Pattern = run {
        val escaped = name.replace(Regex("([\\\\.\\[{(*+?^\$|])"), "\\\\$1")
        Pattern.compile(
            escaped
                .replace("\\*\\*", ".+?")
                .replace("\\*", "[^\\\\]+?")
                .replace("\\?", ".")
                .replace("\\|", "|")
        )
    }

    @set:JvmName("setClassFilter")
    var _classFilter: PointFilter = TruePointFilter

    override fun getClassFilter() = _classFilter

    override fun matches(element: PhpNamedElement): Boolean {
        if (element is PhpClassMember && !modifierFilter.matches(element)) return false

        val elementName = when (element) {
            is PhpClassMember -> element.name
            is PhpClass -> element.fqn?.removePrefix("\\") ?: ""
            else -> ""
        }

        return elementName == name || regexp.matcher(elementName).matches()
    }

    override fun getKind() = filterKind

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SignaturePointcut) return false
        return filterKind == other.filterKind && name == other.name
            && modifierFilter == other.modifierFilter && _classFilter == other._classFilter
    }

    override fun hashCode(): Int {
        var result = filterKind.hashCode()
        result = 31 * result + name.hashCode()
        result = 31 * result + modifierFilter.hashCode()
        result = 31 * result + _classFilter.hashCode()
        return result
    }
}
